qip-0013.md

```diff
@@ -0,0 +1,166 @@
+
+#### Definitions
+* Payment code: an extended public key and associated metadata which is associated with a particular identity/account
+* Mailbox contract: an agreed upon contract where Bob will check for notification messages
+* Notification transaction: a transaction where Alice leaves a notification message (her payment code) in Bob's mailbox.
+* Designated pubkey: the public key necessary for the other party to perform ECDH exchange and derive your addresses using your payment code.
+
+#### Notification Transaction
+Prior to the first time Alice initiates a transaction to Bob, Alice MUST inform Bob of her payment code via the following procedure:
+
+1) Alice constructs a transaction which sends a message to the agreed upon mailbox contract, containing her payment code for Bob.
+1) Alice derives a unique shared secret using ECDH:
+    1) Alice selects the private key corresponding to the her designated pubkey: <pre>a</pre>
+    1) Alice selects Bob's designated public key associated with Bob's payment code: <pre>B, where B = bG</pre>
+    1) Alice calculates a secret point: <pre>S = aB</pre>
+    1) Alice calculates a 64 byte blinding factor: <pre>s = HMAC-SHA512(o, x)</pre>
+        1) "x" is the x value of the secret point
+        1) "o" is the outpoint being spent by the designated input
+1) Alice serializes her payment code in binary form.
+1) Alice renders her payment code (P) unreadable to anyone except Bob:
+    1) Replace the x value with x': <pre>x' = x XOR (first 32 bytes of s)</pre>
+    1) Replace the chain code with c': <pre>c' = c XOR (last 32 bytes of s)</pre>
+1) Alice calls the `notify(...)` message on the agreed upon notification contract.
+
+1) Bob watches the mailbox contract for new notification messages
+1) When a notification is received, if the first byte of the payload in a notification transaction is 0x01:
+    1) Bob selects Alice's designated pubkey from her payment code: <pre>A, where A = aG</pre>
+    1) Bob selects the private key associated with designated pubkey from his payment code: <pre>b</pre>
+    1) Bob calculates a secret point: <pre>S = bA</pre>
+    1) Bob calculates the blinding factor: <pre>s = HMAC-SHA512(x, o)</pre>
+        1) "x" is the x value of the secret point
+        1) "o" is the outpoint being spent by the designated input.
+    1) Bob interprets the 80 byte payload as a payment code, except:
+        1) Replace the x value with x': <pre>x' = x XOR (first 32 bytes of s)</pre>
+        1) Replace the chain code with c': <pre>c' = c XOR (last 32 bytes of s)</pre>
+    1) If the updated x value is a member of the secp256k1 group, the payment code is valid.
+    1) If the updated x value is not a member of the secp256k1 group, the payment code is ignored.
+
+Now that Bob's client has received Alice's payment code, it is possible for Alice to send payments (up to 2<sup>32</sup> payments) to Bob.
+
+Alice will never again need to send a notification transaction to Bob.
+
```

Comment:
> the message to the mailbox contract can happen concurrently with the send of the funds

How do you intend to do this? Our version of the EVM does not support execution batching, so this still requires two TXs to happen, even if they happen back to back.

> bob can provide alice the mailbox contract address at the same time that he provides his address

Bob never shares his address. He shares a payment code with a designated pubkey and a mailbox address. The designated pubkey does not correspond to an address.

> the mailbox contract is not required to be used if alice communicates the information to bob in an alternate format eg a QR code

that is not true. The mailbox contract is always required Mailbox data is necessary for Alice and Bob to recover their wallets. Therefore, I intentionally require the mailbox to be used, so that users may recover their accounts with just a seed phrase.

Importantly, this applies to both parties. If Bob doesn't post his code for whatever business reasons he has, then Alice would not be able to  recover her wallet. The mailbox contract must be used.

> if bob wants to make the channel "persistant" if alice delivers the message in an ephemeral form, he can make the transaction to the mailbox contract himself.

see above. "ephemeral forms" are intentionally not allowed in this spec.

> if bob wants to make the channel "persistant" if alice delivers the message in an ephemeral form, he can make the transaction to the mailbox contract himself.

see above.

> There should be a specification that defines how bob shares this data with alice, specifically, , where endpoint is either a mailbox contract, but alternatively could be a centralized service, server, or could specify "QR" as the mechanism to share the return data.

The community just needs to agree on an address URI encoding (which may additionally be QR encoded), and someone can add a QIP that defines that at a later date.

> It would be helpful, if not here, but somewhere to define the additional functions that will be needed to extend a bip44 wallet to support this.

Not sure what you mean. This QIP defines the methematical operations that need to happen. There's an infinite number of ways someone could choose to write code to do perform these operations.

Reply:
I just pushed some changes to address a few of these. Importantly, Bob never shares his address or a mailbox contract address.
1) Bob shares his designated pubkey (alternatively, he could share his address and some signed message) with anyone who wants to send him money
1) Alice looks up the corresponding contract by deriving an address from that key, determining which shard it belongs to, and looking up the contract address for that shard.
1) Alice builds her payment code for Bob and notifies him in that same contract.